### PR TITLE
Add bind mounts to fuse-wake

### PR DESCRIPTION
**Pull request by mmjconolly**
# Add bind mount support to fuse-wake

The input json has fields added for the mounts.
They must be correctly ordered if there is any overlap and the mount types must be supported by the platform.
The mounts are non-recursive and are optionally read-only when supported by the underlying filesystem.
If at some point we add more mount types, we can add a 'mount-type' field to the json entries

This does not add anything to Plan or Job to use this from within wake.
Jobs that do not use bind mounts are unaffected

This was tested by running `fuse-wake in.json out.json`
This performs a copy of `/bin/bash` into `/home/matthewc/bind-mounts/test/ws-mount2`, except that the copy command uses neither of those two paths directly

`in.json`:
```
{
  "command": [
        "bash", "-c", "ls -l ws-mount*; cp -v ws-mount1/bash ws-mount3/; ls ws-mount2; findmnt"
  ],
  "environment": [
    "PATH=/usr/bin:/bin"
  ],
  "visible": ["ws-mount1", "ws-mount2", "ws-mount3"],
  "directory": ".",
  "mounts": [
    {
      "source": "/bin",
      "destination": "/home/matthewc/bind-mounts/test/ws-mount1",
      "read-only": true
    },
    {
      "source": "/home/matthewc/bind-mounts/test/ws-mount2",
      "destination": "/home/matthewc/bind-mounts/test/ws-mount3"
    }
  ],
  "stdin": "",
  "resources": [],
  "version": "v0.21.0-alpha-4-gde85a32-dirty",
  "usage": {
    "status": 1,
    "runtime": 5.40870000000000031e-02,
    "cputime": 4.00000000000000008e-03,
    "membytes": 3289088,
    "inbytes": 0,
    "outbytes": 0
  }
}

```

### CODE REVIEW

**Comment by mmjconolly** [comment]
Actually, just noticed I wasn't handling errno correctly

File: fuse/fuse.cpp [hunk 1]
```diff
@@ -62,6 +62,45 @@ static void map_id(const char *file, uint32_t from, uint32_t to)
...
+		}
+	}
+	return true;
+}
+
+// Do the mounts specified in the parsed input json.
+// The input/caller responsibility to ensure that the mountpoint exists,
+// that the platform supports the mount type/options, and to correctly order
+// the layered mounts.
+static bool json_mounts(JAST jast)
```
Comment (mmjconolly) [inline fuse/fuse.cpp:91]: const JAST& jast

File: fuse/fuse.cpp [hunk 2]
```diff
@@ -62,6 +62,45 @@ static void map_id(const char *file, uint32_t from, uint32_t to)
...
+}
+
+// Do the mounts specified in the parsed input json.
+// The input/caller responsibility to ensure that the mountpoint exists,
+// that the platform supports the mount type/options, and to correctly order
+// the layered mounts.
+static bool json_mounts(const JAST& jast)
+{
+	for (auto &x : jast.get("mounts").children) {
+		std::string src = x.second.get("source").value;
```
Comment (mmjconolly) [inline fuse/fuse.cpp:94]: const std::string src

**Comment by terpstra** [comment]
Long term filesystem plan?:

```json
"mounts": [
    {
      "source": "images/busybox.squashfs",
      "destination": "/"
    },
    {
      "source": "/sys",
      "destination": "/sys",
      "writeable": true
    },
    {
      "source": "/proc",
      "destination": "/proc",
      "writeable": true
    },
    {
      "source": "/dev",
      "destination": "/dev",
      "writeable": true
    },
    {
      "source": "tmpfs",
      "destination": "/tools",
      "writeable": true
    {
      "source": "./images/java9.squashfs",
      "destination": "/tools/java",
      "mkdir": true
     },
     {
       "source": ".",
       "destination": "/build",
       "writeable": true
      } 
]   
```

**Review by terpstra — APPROVED** [review]
Looks good!

**Pull request merged**